src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs

```diff
@@ -43,124 +38,93 @@ public class CpuUsageExtensionPoint : IDisposable
-        {
-            try
-            {
-                var errors = await _process.StandardError.ReadLineAsync();
-                if (errors == null)
-                {
-                    return;
-                }
-                try
-                {
-                    if (_process.HasExited == false)
-                    {
-                        _process.Kill();
-                    }
-                }
-                catch (Exception)
-                {
-                    // When the process terminating, killed or exited
-                }
-                errors = errors + Environment.NewLine + await _process.StandardError.ReadToEndAsync();
-                NotifyWarning($"Extension point process send an error: {errors}");
-            }
-            catch (Exception e)
-            {
-                NotifyWarning("Could not read errors from cpu usage extension point process", e);
-            }
-        }
 
-        public void Dispose()
-        {
-            IsDisposed = true;
-            using (_process)
-            {
-                try
-                {
-                    if (_process.HasExited == false)
-                    {
-                        _process.Kill();
-                    }
-                }
-                catch (Exception)
-                {
-                    // When the process terminating, killed or exited
-                }
-
-                _process = null;
-            }
-        }
-
-
-        private async Task ReadProcess()
+        private void StartInternal(CancellationToken ctk)
         {
             try
             {
-                while (IsDisposed == false)
+                var retry = 2;
+                const int maxMinutesBetweenIssues = 15;
+                var lastRestart = DateTime.UtcNow;
+                var lastReceivedLine = DateTime.UtcNow;
+                while (retry-- > 0)
                 {
-                    var nextLine = _process.StandardOutput.ReadLineAsync();
-
-                    if (await nextLine.WaitWithTimeout(_timeout) == false)
+                    using (var cts = new CancellationTokenSource())
+                    using (var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, ctk))
                     {
-                        // here we missed an update, restart the process once
-                        HandleError($"The process didn't send information for {_timeout.TotalSeconds} seconds.");
-                        break;
-                    }
-
-                    var line = await nextLine;
-                    if (HandleInfoReceived(line) == false)
-                    {
-                        break;
+                        var lineOutHandler = new Action<object, string>((p, l) =>
+                        {
+                            if (l == null)
+                            {
+                                if (DateTime.UtcNow - lastReceivedLine > TimeSpan.FromSeconds(60))
```

Comment:
Am I reading this wrong, what can cause us to get a `null` here except at the end of the input?

Reply:
`ReadLineAsync` will only return null on end of input, no other scenario

Reply:
True.. (https://docs.microsoft.com/en-us/dotnet/api/system.io.streamreader.readlineasync?view=netframework-4.8#returns)
So we read line by line until no output.. I do not see where the problem is... (?)

Reply:
Your code here does: `if (DateTime.UtcNow - lastReceivedLine`
It is attempting to handle the case where we have no input from the process.
But the current behavior means that it will never happen

Reply:
Ok, you mean ReadLineAsync will block if process hangs and will not return null. 
iirc I've checked that but I will check again.

Reply:
Yes, there is no input there, so it will not return anything.